File: app_for_fluuter/android/app/src/main/java/plugins/ThirdMapPlugin.java
package plugins;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import java.io.File;
import java.util.ArrayList;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.PluginRegistry;



public class ThirdMapPlugin implements MethodCallHandler {

    public static String CHANNEL = "com.zcj.third_map_plugin.cn";

    private Context context;

    public ThirdMapPlugin(Context context) {
        this.context = context;
    }

    public static void registerWith(PluginRegistry.Registrar registrar) {
        final MethodChannel channel =
                new MethodChannel(registrar.messenger(), CHANNEL);
        ThirdMapPlugin instance = new ThirdMapPlugin(registrar.context());
        channel.setMethodCallHandler(instance);
    }

    @Override
    public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {

        switch (methodCall.method) {
            case "hasInstalledMapApp":
                result.success(hasInstalledMapApp());
                break;
            case "openBaiduMap":
                openBaiduMap((Map<String, String>) methodCall.arguments);
                break;
            case "openAmap":
                openAmap((Map<String, String>) methodCall.arguments);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private ArrayList<String> hasInstalledMapApp() {
        ArrayList<String> installedApps = new ArrayList<>();

        // 百度地图包名。/data/data/com.baidu.BaiduMap
        File baiduFile = new File("/data/data/com.baidu.BaiduMap");
        if (baiduFile.exists()) {
            installedApps.add("百度地图");
        }

        // 高德地图包名。 /data/data/com.autonavi.minimap
        File amapFile = new File("/data/data/com.autonavi.minimap");
        if (amapFile.exists()) {
            installedApps.add("高德地图");
        }
        return  installedApps;
    }


    private void openBaiduMap(Map<String, String> argmentsDic) {
        String dlatitude = argmentsDic.get("dlatitude");
        String dlongitude = argmentsDic.get("dlongitude");
        String dName = argmentsDic.get("dName");


        String baiduUrl = "baidumap://map/direction?destination=latlng:" +
                dlatitude + "," +
                dlongitude + "|name:" +
                dName +
                "&src=andr.baidu.openAPIdemo";
        openThirdMap(baiduUrl);
    }

    private void openAmap(Map<String, String> argmentsDic) {
        String dlatitude = argmentsDic.get("dlatitude");
        String dlongitude = argmentsDic.get("dlongitude");
        String dName = argmentsDic.get("dName");


        String amapUrl = "androidamap://route/plan?dlat=" +
                dlatitude + "&dlon=" +
                dlongitude + "&dname=" + dName;

        openThirdMap(amapUrl);
    }

    private void openThirdMap(String uri) {
        Intent intent = new Intent();
        intent.setData(Uri.parse(uri));
        context.startActivity(intent);
    }


}

File: app_for_fluuter/ios/Runner/MethodChannel/ThirdMapPlugin.swift
import Flutter
import UIKit
import MapKit

public class ThirdMapPlugin: NSObject, FlutterPlugin {

  var result: FlutterResult!
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "com.zcj.third_map_plugin.cn", binaryMessenger: registrar.messenger())
    let instance = ThirdMapPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    self.result = result
    
    if(call.method == "getPlatformVersion"){
        result("iOS " + UIDevice.current.systemVersion)
        return
    }
    if(call.method == "hasInstalledMapApp"){
        let installedMaps = hasInstalledMapApp()
        result(installedMaps)
        return
    }
    if(call.method == "openBaiduMap"){
        let argument = call.arguments!
        guard let argumentDic = argument as? Dictionary<String, String> else {
            return
        }
        openBaiduMap(argumentDic: argumentDic)
        return
    }
    if(call.method == "openAmap") {
        let argument = call.arguments!
        guard let argumentDic = argument as? Dictionary<String, String> else {
            return
        }
        openAmap(argumentDic: argumentDic)
        return
    }
    if(call.method == "openAppleMap") {
        let argument = call.arguments!
        guard let argumentDic = argument as? Dictionary<String, String> else {
            return
        }
        openAppleMap(argumentDic: argumentDic)
        return
    }
    
    
    result(FlutterMethodNotImplemented)
  }
    
}

extension ThirdMapPlugin {
    
    private func hasInstalledMapApp() ->Array<String>{
        var results = Array<String>();
        
        let application = UIApplication.shared
        if application.canOpenURL(URL(string: "baidumap://")!) {
            results.append("百度地图")
        }
        if application.canOpenURL(URL(string: "iosamap://")!) {
            results.append("高德地图")
        }
        if application.canOpenURL(URL(string: "http://maps.apple.com/")!) {
            results.append("苹果地图")
        }
        return results
    }
    
    
    //baidumap://map/direction?destination=$dlatitude,$dlongitude&src=andr.baidu.openAPIdemo
    private func openBaiduMap(argumentDic: Dictionary<String, String>) {
        
        let dlongitude = argumentDic["dlongitude"]!
        let dlatitude = argumentDic["dlatitude"]!
        let dName = argumentDic["dName"]!

        let url = "baidumap://map/direction?destination=latlng:\(dlatitude),\(dlongitude)|name:\(dName)&src=andr.baidu.openAPIdemo"
        openThirdMap(url: url)
    }
    
    //iosamap://path?dlat=39.98848272&dlon=116.47560823&dname=B
    private func openAmap(argumentDic: Dictionary<String,String>) {
        
        let dlongitude = argumentDic["dlongitude"]!
        let dlatitude = argumentDic["dlatitude"]!
        let dName = argumentDic["dName"]!
        
        let url = "iosamap://path?dlat=\(dlatitude)&dlon=\(dlongitude)&dname=\(dName)"
        openThirdMap(url: url)
    }
    
    // ios 10.0 之前："http://maps.apple.com/"
    // 10.0之后，通过MapItem打开
    private func openAppleMap(argumentDic: Dictionary<String,String>) {
        
        let dlongitude = Double(argumentDic["dlongitude"]!) ?? 0
        let dlatitude = Double(argumentDic["dlatitude"]!) ?? 0
        let dName = argumentDic["dName"]!
        if #available(iOS 10.0, *) {
            let dcoordinate2d = CLLocationCoordinate2DMake(dlatitude, dlongitude)
            let mapItem = MKMapItem(placemark: MKPlacemark(coordinate: dcoordinate2d))
            mapItem.name = dName
            mapItem.openInMaps(launchOptions: nil)
        } else {
            let url = "http://maps.apple.com/"
            openThirdMap(url: url)
        }
    }
    
    private func openThirdMap(url: String) {
        
        guard let urlPath = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)  else {
            result(false)
            return
        }
        
        let application = UIApplication.shared
        guard let mapUrl = URL(string: urlPath) else {
            result(false)
            return
        }
        
        guard application.canOpenURL(mapUrl) else{
            result(false)
            return
        }
        
        application.openURL(mapUrl)
        result(true)
    }
}
